### PR TITLE
Small refactor to PyTorch model reader to handle tensors received in node's kwargs.

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:
This PR fixes a bug that the "Const Representation" PR reveiled, that input tensors to a node through it's kwargs were not handled correctly.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 1]
```diff
@@ -53,6 +53,25 @@ def extract_holder_weights(constant_name, node_target, model, weights, to_numpy)
...
+def get_node_args_spec(node: Node) -> List[str]:
+    """
+    Return a list of the node's function args & kwargs, so we can insert the kwarg as arg
+    Args:
+        node: Torch.fx.Node to extract target from, which the function the Node executes.
+
+    Returns: A list of args and kwargs names.
+
+    """
+    if node.target in [torch.nn.functional.conv1d, torch.nn.functional.conv2d, torch.nn.functional.conv3d]:
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:65]: why the special treatment for these specific layer types?

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 2]
```diff
@@ -53,6 +53,25 @@ def extract_holder_weights(constant_name, node_target, model, weights, to_numpy)
     return weights
 
 
+def get_node_args_spec(node: Node) -> List[str]:
+    """
+    Return a list of the node's function args & kwargs, so we can insert the kwarg as arg
+    Args:
+        node: Torch.fx.Node to extract target from, which the function the Node executes.
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:60]: what does "which the function the Node executes" mean?

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 3]
```diff
@@ -53,6 +53,25 @@ def extract_holder_weights(constant_name, node_target, model, weights, to_numpy)
     return weights
 
 
+def get_node_args_spec(node: Node) -> List[str]:
+    """
+    Return a list of the node's function args & kwargs, so we can insert the kwarg as arg
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:58]: give some more explanation, maybe give an example of a problematic case.
the terms "function args", and "insert the kwargs as args" are not clear, I know what you mean just because you explained it to me personally.
So elaborate a little more, something like:
"This function builds an ordered (?) list of layer operation arguments. It flattens and combines both the args list and the kwargs dictionary for a single list, specifically, enables to treat the kwargs as arguments. This is important for the case where functional input (or find a better term), which is a dynamic output of another layer's operation (again, fix my explanation if it is not correct), is passed to the layer as a key argument".

Also, "return**s**", and a '.' at the end of a sentence.

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 4]
```diff
@@ -163,10 +182,15 @@ def nodes_builder(model: GraphModule,
                 framework_attr_filtered[k] = v
         framework_attr = framework_attr_filtered
 
-        # filter Nodes from node kwargs, we replace these attributes with nx graph nodes
+        # filter Nodes from node kwargs, we replace these attributes with nx graph nodes. The Node kwargs
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:185]: maybe try to clarify this as well, I'm afraid it is written in a way that only you can understand because you have the explanation in your head.
I think that there is no escape from providing an example.

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 5]
```diff
@@ -178,11 +202,22 @@ def nodes_builder(model: GraphModule,
                                                  node.args[0].op == PLACEHOLDER and node.args[0].meta[TYPE] in (list, tuple))
             tensor_input_index = []
             op_call_args = list(node.args)
+            for arg_name, arg_val in _args_from_kwargs.items():
+                _args = get_node_args_spec(node)
+                if arg_name in _args:
+                    arg_index = _args.index(arg_name)
+                    # TODO: This code verifies the kwargs with Node values that are used as args are in the correct order.
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:209]: what's the "todo"? even if it just so it will be clear what's going on here, add a sentence about how it should be eventually handled

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 6]
```diff
@@ -178,11 +202,22 @@ def nodes_builder(model: GraphModule,
...
+                    # (i.e. check that the kwarg with Node is the first of the kwargs)
+                    if len(op_call_args) != arg_index:
+                        Logger.warning("A kwarg with value of Node should be the first after the arguments. "
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:212]: what is "value of Node"? (I think I know what it means because you should it to me in the debugger, but it is not clear)

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 7]
```diff
@@ -178,11 +202,22 @@ def nodes_builder(model: GraphModule,
...
+                                       "This might fail the model builder later...")
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:213]: This might cause issues when trying to rebuild the model from the graph"

File: tests/pytorch_tests/model_tests/feature_models/const_representation_test.py [hunk 8]
```diff
@@ -101,3 +101,39 @@ def __init__(self, unit_test):
 
     def create_networks(self):
         return ConstRepresentationMultiInputNet()
+
+
+class ConstRepresentationCodeNet(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.conv2d = nn.Conv2d(16, 16, 3, 2, padding=1)
+        # self.bn = torch.nn.BatchNorm2d(16)
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/const_representation_test.py:110]: remove?

File: tests/pytorch_tests/model_tests/feature_models/const_representation_test.py [hunk 9]
```diff
@@ -101,3 +101,39 @@ def __init__(self, unit_test):
...
+        # TODO: Enable this code when PyTorch model reader is fixed for this case. Also, move this to model reader tests.
+        # y = torch.nn.functional.batch_norm(y[:, :, 0, :], self.bn.running_mean, self.bn.running_var,
+        #                                    momentum=0.2, eps=1e-6, bias=self.bn.bias)
+        # y = torch.unsqueeze(y, 2)
+        return torch.add(x, y)
+
+
+class ConstRepresentationCodeTest(ConstRepresentationTest):
+    """
+    This test checks the code representation code in the model reader and builder, not specifically
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/const_representation_test.py:126]: first "code" --> "const"

File: tests/pytorch_tests/model_tests/feature_models/const_representation_test.py [hunk 10]
```diff
@@ -101,3 +101,39 @@ def __init__(self, unit_test):
...
+    with constants in the graph.
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/const_representation_test.py:127]: so explain what it checks (arguments order and parsing?)

File: tests/pytorch_tests/model_tests/test_feature_models_runner.py [hunk 11]
```diff
@@ -252,6 +252,8 @@ def test_const_representation(self):
 
         ConstRepresentationMultiInputTest(self).run_test()
 
+        ConstRepresentationCodeTest(self, None, None).run_test()
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/test_feature_models_runner.py:255]: what are the 2 Nones?

File: tests/pytorch_tests/model_tests/feature_models/const_representation_test.py [hunk 12]
```diff
@@ -101,3 +101,39 @@ def __init__(self, unit_test):
 
     def create_networks(self):
         return ConstRepresentationMultiInputNet()
+
+
+class ConstRepresentationCodeNet(nn.Module):
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/const_representation_test.py:106]: maybe duplicate this test to check different arguments combinations? (namely, args and kwargs, dynamic, etc.)
do you think it would be valuable to test?

**Review by ofirgo — CHANGES_REQUESTED** [review]
mainly documentation comments, some testing comments

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py (hunk 1 above)
Comment (elad-c) [reply model_compression_toolkit/core/pytorch/reader/graph_builders.py:65]: Because `inspect.getfullargspec(node.target)` fails for them.

File: tests/pytorch_tests/model_tests/feature_models/const_representation_test.py (hunk 12 above)
Comment (elad-c) [reply tests/pytorch_tests/model_tests/feature_models/const_representation_test.py:106]: Added functional `Conv2d`, `ConvTranspose2d` & `Linear` to cover all options of `get_node_args_spec`

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py (hunk 1 above)
Comment (elad-c) [reply model_compression_toolkit/core/pytorch/reader/graph_builders.py:65]: only used here, so no need to add code for it.
